File: frontend/src/api/modules/spatial.ts
/**
 * 空间拓扑管理 API
 */
import request from '@/utils/request'
import type { ResponseModel } from './types'

// ==================== 类型定义 ====================

/** 站点 */
export interface Site {
  id: number
  site_code: string
  site_name: string
  address?: string
  description?: string
  created_at: string
  updated_at: string
}

/** 站点创建/更新参数 */
export interface SiteForm {
  site_code: string
  site_name: string
  address?: string
  description?: string
}

/** 楼层 */
export interface Floor {
  id: number
  floor_code: string
  floor_name: string
  site_id: number
  sort_order: number
  created_at: string
  updated_at: string
}

/** 楼层创建/更新参数 */
export interface FloorForm {
  floor_code: string
  floor_name: string
  site_id: number
  sort_order: number
}

/** 房间 */
export interface Room {
  id: number
  room_code: string
  room_name: string
  floor_id: number
  grid_cols: number
  grid_rows: number
  area_sqm?: number
  description?: string
  created_at: string
  updated_at: string
}

/** 房间创建/更新参数 */
export interface RoomForm {
  room_code: string
  room_name: string
  floor_id: number
  grid_cols: number
  grid_rows: number
  area_sqm?: number
  description?: string
}

/** 列 */
export interface Row {
  id: number
  row_code: string
  row_name: string
  room_id: number
  aisle_type: string
  sort_order: number
  created_at: string
  updated_at: string
}

/** 列创建/更新参数 */
export interface RowForm {
  row_code: string
  row_name: string
  room_id: number
  aisle_type: string
  sort_order: number
}

/** 机柜树节点 */
export interface TreeCabinet {
  id: number
  cabinet_code: string
  cabinet_name: string
  grid_x?: number
  grid_y?: number
  aisle_type?: string
}

/** 列树节点 */
export interface TreeRow {
  id: number
  row_code: string
  row_name: string
  aisle_type: string
  cabinets: TreeCabinet[]
}

/** 房间树节点 */
export interface TreeRoom {
  id: number
  room_code: string
  room_name: string
  grid_cols: number
  grid_rows: number
  rows: TreeRow[]
}

/** 楼层树节点 */
export interface TreeFloor {
  id: number
  floor_code: string
  floor_name: string
  rooms: TreeRoom[]
}

/** 空间拓扑树节点 */
export interface SpatialTreeNode {
  id: number
  site_code: string
  site_name: string
  floors: TreeFloor[]
}

/** 布局模板 */
export interface LayoutTemplate {
  id: number
  template_code: string
  template_name: string
  description?: string
}

/** 导入结果 */
export interface ImportResult {
  total: number
  success: number
  failed: number
  skipped: number
  errors: string[]
}

/** 模板应用结果 */
export interface TemplateApplyResult {
  created_rows: number
  created_cabinets: number
  skipped_cabinets: number
  errors: string[]
}

// ==================== 站点 API ====================

/** 获取站点列表 */
export function getSites(keyword?: string) {
  return request.get<ResponseModel<Site[]>>('/v1/spatial/sites', { params: { keyword } })
}

/** 创建站点 */
export function createSite(data: SiteForm) {
  return request.post<ResponseModel<Site>>('/v1/spatial/sites', data)
}

/** 更新站点 */
export function updateSite(id: number, data: Partial<SiteForm>) {
  return request.put<ResponseModel<Site>>(`/v1/spatial/sites/${id}`, data)
}

/** 删除站点 */
export function deleteSite(id: number) {
  return request.delete<ResponseModel>(`/v1/spatial/sites/${id}`)
}

// ==================== 楼层 API ====================

/** 获取楼层列表 */
export function getFloors(site_id?: number) {
  return request.get<ResponseModel<Floor[]>>('/v1/spatial/floors', { params: { site_id } })
}

/** 创建楼层 */
export function createFloor(data: FloorForm) {
  return request.post<ResponseModel<Floor>>('/v1/spatial/floors', data)
}

/** 更新楼层 */
export function updateFloor(id: number, data: Partial<FloorForm>) {
  return request.put<ResponseModel<Floor>>(`/v1/spatial/floors/${id}`, data)
}

/** 删除楼层 */
export function deleteFloor(id: number) {
  return request.delete<ResponseModel>(`/v1/spatial/floors/${id}`)
}

// ==================== 房间 API ====================

/** 获取房间列表 */
export function getRooms(floor_id?: number) {
  return request.get<ResponseModel<Room[]>>('/v1/spatial/rooms', { params: { floor_id } })
}

/** 创建房间 */
export function createRoom(data: RoomForm) {
  return request.post<ResponseModel<Room>>('/v1/spatial/rooms', data)
}

/** 更新房间 */
export function updateRoom(id: number, data: Partial<RoomForm>) {
  return request.put<ResponseModel<Room>>(`/v1/spatial/rooms/${id}`, data)
}

/** 删除房间 */
export function deleteRoom(id: number) {
  return request.delete<ResponseModel>(`/v1/spatial/rooms/${id}`)
}

// ==================== 列 API ====================

/** 获取列列表 */
export function getRows(room_id?: number) {
  return request.get<ResponseModel<Row[]>>('/v1/spatial/rows', { params: { room_id } })
}

/** 创建列 */
export function createRow(data: RowForm) {
  return request.post<ResponseModel<Row>>('/v1/spatial/rows', data)
}

/** 更新列 */
export function updateRow(id: number, data: Partial<RowForm>) {
  return request.put<ResponseModel<Row>>(`/v1/spatial/rows/${id}`, data)
}

/** 删除列 */
export function deleteRow(id: number) {
  return request.delete<ResponseModel>(`/v1/spatial/rows/${id}`)
}

// ==================== 拓扑树 API ====================

/** 获取空间拓扑树 */
export function getSpatialTree() {
  return request.get<ResponseModel<SpatialTreeNode[]>>('/v1/spatial/tree')
}

// ==================== 机柜位置 API ====================

/** 更新机柜位置 */
export function updateCabinetPosition(cabinetId: number, data: { grid_x: number; grid_y: number }) {
  return request.put<ResponseModel>(`/v1/spatial/cabinets/${cabinetId}/position`, data)
}

// ==================== 导入导出 API ====================

/** 导入空间拓扑 Excel */
export function importSpatialExcel(file: File) {
  const formData = new FormData()
  formData.append('file', file)
  return request.post<ResponseModel<ImportResult>>('/v1/spatial/import', formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}

/** 导出空间拓扑 Excel */
export function exportSpatialExcel() {
  return request.get('/v1/spatial/export', { responseType: 'blob' })
}

// ==================== 模板 API ====================

/** 获取布局模板列表 */
export function getTemplates() {
  return request.get<ResponseModel<LayoutTemplate[]>>('/v1/spatial/templates')
}

/** 应用布局模板 */
export function applyTemplate(templateId: number, data: { room_id: number }) {
  return request.post<ResponseModel<TemplateApplyResult>>(`/v1/spatial/templates/${templateId}/apply`, data)
}

File: backend/app/schemas/spatial.py
"""
空间拓扑数据模型
"""
from typing import Optional, List
from pydantic import BaseModel, Field
from datetime import datetime


# ==================== Site Schemas ====================

class SiteCreate(BaseModel):
    """创建站点"""
    site_code: str = Field(..., description="站点编码")
    site_name: str = Field(..., description="站点名称")
    address: Optional[str] = Field(None, description="地址")
    contact_person: Optional[str] = Field(None, description="联系人")
    contact_phone: Optional[str] = Field(None, description="联系电话")
    contact_email: Optional[str] = Field(None, description="联系邮箱")
    network_config: Optional[dict] = Field(None, description="网络配置(VPN/专线信息)")
    description: Optional[str] = Field(None, description="描述")


class SiteUpdate(BaseModel):
    """更新站点"""
    site_code: Optional[str] = Field(None, description="站点编码")
    site_name: Optional[str] = Field(None, description="站点名称")
    address: Optional[str] = Field(None, description="地址")
    contact_person: Optional[str] = Field(None, description="联系人")
    contact_phone: Optional[str] = Field(None, description="联系电话")
    contact_email: Optional[str] = Field(None, description="联系邮箱")
    network_config: Optional[dict] = Field(None, description="网络配置(VPN/专线信息)")
    description: Optional[str] = Field(None, description="描述")


class SiteResponse(BaseModel):
    """站点响应"""
    id: int = Field(..., description="站点ID")
    site_code: str = Field(..., description="站点编码")
    site_name: str = Field(..., description="站点名称")
    address: Optional[str] = Field(None, description="地址")
    contact_person: Optional[str] = Field(None, description="联系人")
    contact_phone: Optional[str] = Field(None, description="联系电话")
    contact_email: Optional[str] = Field(None, description="联系邮箱")
    network_config: Optional[dict] = Field(None, description="网络配置")
    status: str = Field("active", description="状态")
    description: Optional[str] = Field(None, description="描述")
    gateway_count: int = Field(0, description="网关数量")
    device_count: int = Field(0, description="设备数量")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True


class SiteSummaryItem(BaseModel):
    """站点汇总项"""
    id: int = Field(..., description="站点ID")
    site_code: str = Field(..., description="站点编码")
    site_name: str = Field(..., description="站点名称")
    status: str = Field("active", description="状态")
    gateway_count: int = Field(0, description="网关数量")
    device_count: int = Field(0, description="设备数量")
    active_alarm_count: int = Field(0, description="活跃告警数")


class SiteSummaryResponse(BaseModel):
    """跨站点汇总响应"""
    total_sites: int = Field(0, description="站点总数")
    total_gateways: int = Field(0, description="网关总数")
    total_devices: int = Field(0, description="设备总数")
    total_alarms: int = Field(0, description="活跃告警总数")
    sites: List[SiteSummaryItem] = Field(default_factory=list, description="各站点摘要")


# ==================== Floor Schemas ====================

class FloorCreate(BaseModel):
    """创建楼层"""
    floor_code: str = Field(..., description="楼层编码")
    floor_name: str = Field(..., description="楼层名称")
    site_id: int = Field(..., description="所属站点ID")
    sort_order: int = Field(0, description="排序")


class FloorUpdate(BaseModel):
    """更新楼层"""
    floor_code: Optional[str] = Field(None, description="楼层编码")
    floor_name: Optional[str] = Field(None, description="楼层名称")
    site_id: Optional[int] = Field(None, description="所属站点ID")
    sort_order: Optional[int] = Field(None, description="排序")


class FloorResponse(BaseModel):
    """楼层响应"""
    id: int = Field(..., description="楼层ID")
    floor_code: str = Field(..., description="楼层编码")
    floor_name: str = Field(..., description="楼层名称")
    site_id: int = Field(..., description="所属站点ID")
    sort_order: int = Field(0, description="排序")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True


# ==================== Room Schemas ====================

class RoomCreate(BaseModel):
    """创建房间"""
    room_code: str = Field(..., description="房间编码")
    room_name: str = Field(..., description="房间名称")
    floor_id: int = Field(..., description="所属楼层ID")
    grid_cols: int = Field(20, description="网格列数")
    grid_rows: int = Field(20, description="网格行数")
    area_sqm: Optional[float] = Field(None, description="面积(平方米)")
    description: Optional[str] = Field(None, description="描述")


class RoomUpdate(BaseModel):
    """更新房间"""
    room_code: Optional[str] = Field(None, description="房间编码")
    room_name: Optional[str] = Field(None, description="房间名称")
    floor_id: Optional[int] = Field(None, description="所属楼层ID")
    grid_cols: Optional[int] = Field(None, description="网格列数")
    grid_rows: Optional[int] = Field(None, description="网格行数")
    area_sqm: Optional[float] = Field(None, description="面积(平方米)")
    description: Optional[str] = Field(None, description="描述")


class RoomResponse(BaseModel):
    """房间响应"""
    id: int = Field(..., description="房间ID")
    room_code: str = Field(..., description="房间编码")
    room_name: str = Field(..., description="房间名称")
    floor_id: int = Field(..., description="所属楼层ID")
    grid_cols: int = Field(20, description="网格列数")
    grid_rows: int = Field(20, description="网格行数")
    area_sqm: Optional[float] = Field(None, description="面积(平方米)")
    description: Optional[str] = Field(None, description="描述")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True


# ==================== Row Schemas ====================

class RowCreate(BaseModel):
    """创建行"""
    row_code: str = Field(..., description="行编码")
    row_name: str = Field(..., description="行名称")
    room_id: int = Field(..., description="所属房间ID")
    aisle_type: str = Field("none", description="通道类型: cold/hot/none")
    sort_order: int = Field(0, description="排序")


class RowUpdate(BaseModel):
    """更新行"""
    row_code: Optional[str] = Field(None, description="行编码")
    row_name: Optional[str] = Field(None, description="行名称")
    room_id: Optional[int] = Field(None, description="所属房间ID")
    aisle_type: Optional[str] = Field(None, description="通道类型: cold/hot/none")
    sort_order: Optional[int] = Field(None, description="排序")


class RowResponse(BaseModel):
    """行响应"""
    id: int = Field(..., description="行ID")
    row_code: str = Field(..., description="行编码")
    row_name: str = Field(..., description="行名称")
    room_id: int = Field(..., description="所属房间ID")
    aisle_type: str = Field("none", description="通道类型: cold/hot/none")
    sort_order: int = Field(0, description="排序")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    class Config:
        from_attributes = True


# ==================== LayoutTemplate Schemas ====================

class LayoutTemplateResponse(BaseModel):
    """布局模板响应"""
    id: int = Field(..., description="模板ID")
    template_code: str = Field(..., description="模板编码")
    template_name: str = Field(..., description="模板名称")
    description: Optional[str] = Field(None, description="描述")
    template_data: Optional[str] = Field(None, description="JSON格式模板数据")
    created_at: datetime = Field(..., description="创建时间")

    class Config:
        from_attributes = True


# ==================== Cabinet Position ====================

class CabinetPositionUpdate(BaseModel):
    """更新机柜空间位置"""
    row_id: Optional[int] = Field(None, description="所属行ID")
    aisle_type: Optional[str] = Field(None, description="通道类型: cold/hot/none")
    grid_x: Optional[int] = Field(None, description="网格X坐标")
    grid_y: Optional[int] = Field(None, description="网格Y坐标")


# ==================== Tree Response ====================

class TreeCabinetItem(BaseModel):
    """树形结构中的机柜简要信息"""
    id: int
    cabinet_code: str
    cabinet_name: str
    aisle_type: Optional[str] = None
    grid_x: Optional[int] = None
    grid_y: Optional[int] = None

    class Config:
        from_attributes = True


class TreeRowItem(BaseModel):
    """树形结构中的行"""
    id: int
    row_code: str
    row_name: str
    aisle_type: str = "none"
    cabinets: List[TreeCabinetItem] = []

    class Config:
        from_attributes = True


class TreeRoomItem(BaseModel):
    """树形结构中的房间"""
    id: int
    room_code: str
    room_name: str
    grid_cols: int = 20
    grid_rows: int = 20
    rows: List[TreeRowItem] = []

    class Config:
        from_attributes = True


class TreeFloorItem(BaseModel):
    """树形结构中的楼层"""
    id: int
    floor_code: str
    floor_name: str
    sort_order: int = 0
    rooms: List[TreeRoomItem] = []

    class Config:
        from_attributes = True


class SpatialTreeResponse(BaseModel):
    """空间拓扑树形响应"""
    id: int
    site_code: str
    site_name: str
    floors: List[TreeFloorItem] = []

    class Config:
        from_attributes = True


# ==================== Import/Export ====================

class ImportResultResponse(BaseModel):
    """导入结果响应"""
    total: int = Field(0, description="总行数")
    success: int = Field(0, description="成功数")
    failed: int = Field(0, description="失败数")
    skipped: int = Field(0, description="跳过数")
    errors: List[str] = Field(default_factory=list, description="错误信息列表")


# ==================== Template Apply ====================

class TemplateApplyRequest(BaseModel):
    """应用模板请求"""
    room_id: int = Field(..., description="目标房间ID")
    cabinet_prefix: Optional[str] = Field(None, description="机柜编码前缀")


class TemplateApplyResponse(BaseModel):
    """应用模板响应"""
    created_rows: int = Field(0, description="创建的行数")
    created_cabinets: int = Field(0, description="创建的机柜数")
    skipped_cabinets: int = Field(0, description="跳过的机柜数(编码冲突)")
    errors: List[str] = Field(default_factory=list, description="错误信息列表")
